src/main/kotlin/com/hj/leetcode/kotlin/problem1372/Solution.kt

```diff
@@ -0,0 +1,31 @@
+package com.hj.leetcode.kotlin.problem1372
+
+import com.hj.leetcode.kotlin.common.model.TreeNode
+
+/**
+ * LeetCode page: [1372. Longest ZigZag Path in a Binary Tree](https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/);
+ */
+class Solution {
+    /* Complexity:
+     * Time O(N) and Space O(H) where N and H are the number of nodes and height of root
+     */
+    fun longestZigZag(root: TreeNode?): Int {
+        var longestZigZag = 0
+        dfsZigZag(root, 0, 0) { maxLengthToNode ->
+            if (maxLengthToNode > longestZigZag) longestZigZag = maxLengthToNode
+        }
+        return longestZigZag
+    }
+
+    private fun dfsZigZag(
+        root: TreeNode?,
+        lengthFromLeftEdge: Int,
+        lengthFromRightEdge: Int,
+        sideEffect: (maxLengthToNode: Int) -> Unit
```

Comment:
`sideEffect` is not clear when the function is invoked, `onEachNode` may be a clearer name

Reply:
resolved at edea3d6d338592b3a470993cb0ecc1a07e289107